[GStreamer][WebRTC] Capabilities probing support

philn opened the pull request:
#### ea06abc8501c66cf264414965269f23deab87741
<pre>
[GStreamer][WebRTC] Capabilities probing support
<a href="https://bugs.webkit.org/show_bug.cgi?id=243723">https://bugs.webkit.org/show_bug.cgi?id=243723</a>

Reviewed by Xabier Rodriguez-Calvar.

The GStreamerRegistryScanner can now probe for RTP payloads, depayloaders and extensions, so that
the WebRTC MediaCapabilities now correspond to the ones available on the GStreamer runtime. The
PeerConnectionBackend RTP sender and receiver capabilies also make use of the probed capabilies. The
corresponding code is no longer specific to the LibWebRTC backend, since we use the new
WebRTCProvider base class.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/media-capabilities/decodingInfo.webrtc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/media-capabilities/encodingInfo.webrtc-expected.txt:
* LayoutTests/platform/glib/webrtc/vp9-expected.txt: Added.
* LayoutTests/webrtc/vp9-profile2.html:
* Source/WebCore/Modules/mediacapabilities/MediaCapabilities.cpp:
(WebCore::gatherDecodingInfo):
(WebCore::gatherEncodingInfo):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::receiverCapabilities):
(WebCore::PeerConnectionBackend::senderCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::gstreamerRtpCapatiblities): Deleted.
(WebCore::PeerConnectionBackend::receiverCapabilities): Deleted.
(WebCore::PeerConnectionBackend::senderCapabilities): Deleted.
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::receiverCapabilities): Deleted.
(WebCore::PeerConnectionBackend::senderCapabilities): Deleted.
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr&lt;GstRTPHeaderExtension&gt;):
(WTF::derefGPtr&lt;GstRTPHeaderExtension&gt;):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::ElementFactories::ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::~ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::elementFactoryTypeToString):
(WebCore::GStreamerRegistryScanner::ElementFactories::factory const):
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForMediaType const):
(WebCore::GStreamerRegistryScanner::audioRtpCapabilities):
(WebCore::GStreamerRegistryScanner::videoRtpCapabilities):
(WebCore::probeRtpExtensions):
(WebCore::GStreamerRegistryScanner::fillAudioRtpCapabilities):
(WebCore::GStreamerRegistryScanner::fillVideoRtpCapabilities):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCProvider.cpp:
(WebCore::GStreamerWebRTCProvider::receiverCapabilities):
(WebCore::GStreamerWebRTCProvider::senderCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeAudioEncodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeVideoEncodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeAudioDecodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeVideoDecodingCapabilities):

Canonical link: <a href="https://commits.webkit.org/253297@main">https://commits.webkit.org/253297@main</a>
</pre>

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -721,5 +739,105 @@ GStreamerRegistryScanner::RegistryLookupResult GStreamerRegistryScanner::isConfi
+    }
+
+    // TODO: Probe for video/H265 capabilies.
+    // TODO: Probe for video/AV1 capabilies.
```
IIRC, the styles we can only have FIXMEs.

philn replied:
FIXME it is then

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
```diff
@@ -28,6 +28,7 @@
 #endif
 
 #if USE(GSTREAMER_WEBRTC)
+#include <gst/rtp/gstrtphdrext.h>
```
Why not just <gst/rtp/rtp.h> ?

philn replied:
That's what LSP inserted for me and I didn't complain :) We don't need the full and big rtp includes here after all.

ceyusa replied:
AFAIK the grouping headers are preferred by GStreamer, I guess because possible changes in the future. But perhaps it's just bikesheding :)

philn replied:
let's do the right thing then!

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -30,6 +30,10 @@
 #include <wtf/text/StringToIntegerConversion.h>
 
+#if USE(GSTREAMER_WEBRTC)
+#include <gst/rtp/gstrtphdrext.h>
```
ditto

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -721,5 +739,105 @@ GStreamerRegistryScanner::RegistryLookupResult GStreamerRegistryScanner::isConfi
+        capabilities.codecs.append({ .mimeType = "video/VP9"_s, .clockRate = 90000, .sdpFmtpLine = "profile-id=0"_s });
+        capabilities.codecs.append({ .mimeType = "video/VP9"_s, .clockRate = 90000, .sdpFmtpLine = "profile-id=2"_s });
+    }
+}
```
But these capabilities would also depend of the available decoders and encoders in the gstreamer setup, isn't it?

philn replied:
Yes, the scanner already handles that. Not sure what you think is missing here :)

ceyusa replied:
Oh. Then I would be worried about the available profiles. Does the scanner also check that? But I don't want to bikeshed again

philn replied:
There is some support for checking H.264 profiles/levels but it's not hooked in the WebRTC capabilities probing yet, there is a TODO (or FIXME, whatever) about this.

philn committed:
[GStreamer][WebRTC] Capabilities probing support
https://bugs.webkit.org/show_bug.cgi?id=243723

Reviewed by Xabier Rodriguez-Calvar.

The GStreamerRegistryScanner can now probe for RTP payloads, depayloaders and extensions, so that
the WebRTC MediaCapabilities now correspond to the ones available on the GStreamer runtime. The
PeerConnectionBackend RTP sender and receiver capabilies also make use of the probed capabilies. The
corresponding code is no longer specific to the LibWebRTC backend, since we use the new
WebRTCProvider base class.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/media-capabilities/decodingInfo.webrtc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/media-capabilities/encodingInfo.webrtc-expected.txt:
* LayoutTests/platform/glib/webrtc/vp9-expected.txt: Added.
* LayoutTests/webrtc/vp9-profile2.html:
* Source/WebCore/Modules/mediacapabilities/MediaCapabilities.cpp:
(WebCore::gatherDecodingInfo):
(WebCore::gatherEncodingInfo):
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::receiverCapabilities):
(WebCore::PeerConnectionBackend::senderCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::gstreamerRtpCapatiblities): Deleted.
(WebCore::PeerConnectionBackend::receiverCapabilities): Deleted.
(WebCore::PeerConnectionBackend::senderCapabilities): Deleted.
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::PeerConnectionBackend::receiverCapabilities): Deleted.
(WebCore::PeerConnectionBackend::senderCapabilities): Deleted.
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr<GstRTPHeaderExtension>):
(WTF::derefGPtr<GstRTPHeaderExtension>):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::ElementFactories::ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::~ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::elementFactoryTypeToString):
(WebCore::GStreamerRegistryScanner::ElementFactories::factory const):
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForMediaType const):
(WebCore::GStreamerRegistryScanner::audioRtpCapabilities):
(WebCore::GStreamerRegistryScanner::videoRtpCapabilities):
(WebCore::probeRtpExtensions):
(WebCore::GStreamerRegistryScanner::fillAudioRtpCapabilities):
(WebCore::GStreamerRegistryScanner::fillVideoRtpCapabilities):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerWebRTCProvider.cpp:
(WebCore::GStreamerWebRTCProvider::receiverCapabilities):
(WebCore::GStreamerWebRTCProvider::senderCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeAudioEncodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeVideoEncodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeAudioDecodingCapabilities):
(WebCore::GStreamerWebRTCProvider::initializeVideoDecodingCapabilities):

Canonical link: https://commits.webkit.org/253297@main
